### PR TITLE
Fix corrupt floodfill

### PR DESCRIPTION
**Pull request by FirefoxMetzger**
## Description

Closes: https://github.com/scikit-image/scikit-image/issues/4918

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Review by sciunto — APPROVED** [review]

File: skimage/morphology/_flood_fill.py [hunk 1]
```diff
@@ -236,13 +234,20 @@ def flood(image, seed_point, *, selem=None, connectivity=None, tolerance=None):
     except TypeError:
         seed_point = (seed_point,)
 
+    seed_point_as_array = np.asarray_chkfinite(seed_point)
+    if not np.all(
+            (0 <= seed_point_as_array) & (seed_point_as_array < image.shape)):
+        raise IndexError("seed_point lies outside the image.")
```
Comment (jni) [inline skimage/morphology/_flood_fill.py:240]: @FirefoxMetzger I'd prefer to give users the convenience of using negative indexing. Additionally, I don't think `chkfinite` is needed since the indices should be int anyway. I'd suggest:

```suggestion
    seed_point = tuple(np.asarray(seed_point) % working_image.shape)
```

... But that needs to go after the `working_image` definition, below.

Comment (FirefoxMetzger) [reply skimage/morphology/_flood_fill.py:240]: @jni I saw the suggestion in the issue. I think it's a bit strange to use `%`, because it allows you to specify numbers completely outside the image like `seed_point = 2.5 * np.asarray(image.shape, dtype=np.intp)`, but why not.

Comment (FirefoxMetzger) [reply skimage/morphology/_flood_fill.py:240]: Oh, also shouldn't it be `image.shape` instead of `working_image.shape` as we want to avoid the seed being located on the padding, which is what causes the segfault in the first place?

Comment (jni) [reply skimage/morphology/_flood_fill.py]: > also shouldn't it be image.shape instead of working_image.shape as we want to avoid the seed being located on the padding, which is what causes the segfault in the first place?

... 🤦 of course! Well spotted. =)

> I think it's a bit strange to use %, because it allows you to specify numbers completely outside the image like `seed_point = 2.5 * np.asarray(image.shape, dtype=np.intp)`

Sure, and for those cases I wouldn't be opposed to a check that ensures that the seed is in `[-image.shape, image.shape)`, but I think that's a rather minor concern. I generally think undocumented behaviour is ok, as long as the failure mode is not catastrophic (such as a segfault! 😂).

File: skimage/morphology/_flood_fill.py [hunk 2]
```diff
@@ -236,13 +234,16 @@ def flood(image, seed_point, *, selem=None, connectivity=None, tolerance=None):
     except TypeError:
         seed_point = (seed_point,)
 
+    seed_value = image[seed_point]
```
Comment (jni) [inline skimage/morphology/_flood_fill.py:237]: @FirefoxMetzger actually, this will fail with out-of-range seed points, so I think we're covered. =)

File: skimage/morphology/tests/test_flood_fill.py [hunk 3]
```diff
@@ -219,7 +219,7 @@ def test_selem():
                          [0, 255, 255, 255, 255, 255],
                          [0, 255, 255, 255, 255, 255],
                          [0, 255, 255, 255, 255, 255],
-                         [0,   0,   0,   0,   0,   0]], dtype=np.uint8)
+                         [0, 0, 0, 0, 0, 0]], dtype=np.uint8)
```
Comment (jni) [inline skimage/morphology/tests/test_flood_fill.py:222]: Could you please revert this change?

Comment (jni) [reply skimage/morphology/tests/test_flood_fill.py:222]:
```suggestion
                         [0,   0,   0,   0,   0,   0]], dtype=np.uint8)
```

File: skimage/morphology/tests/test_flood_fill.py [hunk 4]
```diff
@@ -230,11 +230,11 @@ def test_selem():
     output = flood_fill(np.zeros((5, 6), dtype=np.uint8), (1, 4), 255,
                         selem=selem)
 
-    expected = np.array([[  0,   0,   0,   0,   0,   0],
-                         [255, 255, 255, 255, 255,   0],
-                         [255, 255, 255, 255, 255,   0],
-                         [255, 255, 255, 255, 255,   0],
-                         [255, 255, 255, 255, 255,   0]], dtype=np.uint8)
```
Comment (jni) [inline skimage/morphology/tests/test_flood_fill.py:237]: And here. (Sorry, this suggestion is harder to do on the web interface 😬)

File: skimage/morphology/tests/test_flood_fill.py [hunk 5]
```diff
@@ -275,5 +275,23 @@ def test_f_order(tolerance):
     np.testing.assert_array_equal(expected, mask)
 
 
+def test_out_of_bounds_seed_point():
```
Comment (jni) [inline skimage/morphology/tests/test_flood_fill.py:278]: I'd rename this test, `test_negative_indexing_seed_point`

**Review by jni — APPROVED** [review]
@FirefoxMetzger other than some very minor formatting/naming suggestions, this is ready for me!

The Travis failures appear unrelated. One is a sporadic failure in the [rank filters tests](https://travis-ci.org/github/scikit-image/scikit-image/jobs/724676450#L4318-L4458), the other is the [mac build hanging](https://travis-ci.org/github/scikit-image/scikit-image/jobs/724676453) which we have encountered elsewhere.

**Comment by emmanuelle** [comment]
Also approving once's @jni's small comments have been addressed.

**Comment by FirefoxMetzger** [comment]
@jni @emmanuelle done. Let's wait for the tests to pass and then this *should* be good.

**Comment by FirefoxMetzger** [comment]
@jni LGTM

**Comment by emmanuelle** [comment]
Thank you @FirefoxMetzger , merging!

**Comment by jni** [comment]
@meeseeksdev backport to v0.17.x

**Comment by jni** [comment]
Dang. LOL
